### ISSUE TITLE
On update URI::InvalidURIError: URI must be ascii only 

### ISSUE THREAD

**Issue by luckyraul** [visoft/ruby_odata#62]
```
conn = OData::Service.new "same_ip/hamster/odata/standard.odata"
conn.Document_СчетНаОплатуПокупателю
results = conn.execute
doc = results.first
doc.Комментарий = 'ruby update'

conn.update_object(doc)
result = conn.save_changes
puts "Was updated? #{result}"
```

**Comment by visoft** [comment]
Try the `develop` branch. I believe I fixed this.

**Comment by luckyraul** [comment]
Uri error is gone, however 
```
OData::ServiceError: XML parsing error:  - [1,1]
Fatal error: 
Document is empty
```

**Comment by visoft** [comment]
I got that too, I figured I didn't have access to update your entities. 

I'll look into this.

**Comment by luckyraul** [comment]
What was the server answer? This service is not our application, just some vendor server software, so there can be some restrictions on editing, which I can't predict. If you find some restrictions responses - I can look in configuration.

**Comment by luckyraul** [comment]
Yeah. As far as I am concerned it does not support json yet.

**Comment by luckyraul** [comment]
I updated service. So It you want to json answer you should request with header ```Accept: application/json``` or parameter ```?$format=json```

**Comment by luckyraul** [comment]
I dumped http and found that library use always "accept xml"

**Comment by visoft** [comment]
Yes, it always accepts XML, but when it sends out say an update, it's content-type is set to `application/json`

**Comment by visoft** [comment]
`ruby_odata` doesn't want JSON back (`Accept`) otherwise I'd have to have 2 different parsers

**Comment by luckyraul** [comment]
here said to use PATCH method http://www.odata.org/getting-started/basic-tutorial/

**Comment by visoft** [comment]
> The OData services SHOULD support PATCH as the preferred means of updating an entity. But also services MAY additionally support PUT.

`ruby_odata` uses a `PUT` to send over the updates since it's updating the whole entity, not just pieces of it (which is what `PATCH`should be used for).

**Comment by luckyraul** [comment]
I've checked my issue and added ?$format=json to url and got error answer
```
{
"odata.error": {
"code": "4",
"message": {
"lang": "ru",
"value": "ВозвратнаяТара"
}
}
``

**Comment by luckyraul** [comment]
so can you tell me how can I use xml for saving?

**Comment by visoft** [comment]
It would require rewriting the save logic

**Comment by luckyraul** [comment]
So if I save with ?format=json I got error an eror with property and json as a result. I guess patching object will be better for my problem. Or try xml.

**Comment by luckyraul** [comment]
I spent some time looking in your's code. Implementing XML saving will be a huge amount of work. 
Now I am trying now use json saving with $format=json parameter

this patch sample from documentation
```
PATCH http://host/base/odata/standard.odata/Catalog_Магазины(guid'value')?$format=json HTTP/1.1
Host: host
Connection: keep-alive
Accept: application/json
Content-Length: 638
{
    "odata.metadata": "http://host/base/odata/standard.odata/$metadata#Catalog_Магазины/@Element",
    "Description": "Новое описание магазина",
    "ТорговыеЗалы@odata.type" : "Collection(StandardODATA.Catalog_Магазины_ТорговыеЗалы_RowType)"
    "ТорговыеЗалы": [
        {
            "LineNumber": "1",
            "Название": "Синий зал",
            "Площадь": 56,
            "ДатаОткрытия": "2015-01-01T00:00:00"
    },
    {
            "LineNumber": "2",
            "Название": "Красный зал",
            "Площадь": 56,
            "ДатаОткрытия": "2015-06-13T11:45:41"
    }
    ]
}
```

In my case ruby is sending 
```
"ВозвратнаяТара": {
    "__metadata": {
      "type": "Collection(StandardODATA.Document_СчетНаОплатуПокупателю_ВозвратнаяТара_RowType)"
    },
    "results": [

    ]
  },
  "УдалитьУслуги": {
    "__metadata": {
      "type": "Collection(StandardODATA.Document_СчетНаОплатуПокупателю_УдалитьУслуги_RowType)"
    },
    "results": [

    ]
  }
```

and server responded error

```
{
  "odata.error": {
    "code": "4",
    "message": {
      "lang": "ru",
      "value": "ВозвратнаяТара"
    }
  }
}

```
which means that problem is in the Collections. 
Error 4 means - Invalid property value

**Comment by luckyraul** [comment]
Do you have any comments?

**Comment by visoft** [comment]
Are you changing anything before saving? The empy `results` collection seems like nothing has changed.

**Comment by visoft** [comment]
Also, the fix I put into `develop` to fix the ASCII problem, doesn't work properly (failed tests), so I need to revisit that problem.

**Comment by luckyraul** [comment]
```doc.Комментарий = 'ruby update'```

I trying to point to the difference beetween '__metadata' inside collection object and property "property@odata.type"

**Comment by luckyraul** [comment]
Let me check

**Comment by visoft** [comment]
Do you know the version of your OData service?

**Comment by luckyraul** [comment]
in metadata - v3

**Comment by luckyraul** [comment]
I've looked in standard reference @odata.type is for 4, __metadata for 3

**Comment by luckyraul** [comment]
in documentation they said it is 3 version. So I don't understand what's wrong with collection object

**Comment by visoft** [comment]
Your service comes back with:

    <edmx:DataServices xmlns:m="http://schemas.microsoft.com/ado/2007/08/dataservices/metadata" m:DataServiceVersion="3.0" m:MaxDataServiceVersion="3.0">

While the Northwind sample comes back with:

    <edmx:DataServices xmlns:m="http://schemas.microsoft.com/ado/2007/08/dataservices/metadata" m:DataServiceVersion="1.0" m:MaxDataServiceVersion="3.0">

I'll bet something changed in v3 and because it only supports 3, the ruby_odata code must not be sending the correct thing. I'll have to look at it more closely this weekend (hopefully I will have time).

It's weird that queries still work though, I guess they didn't change that spec.

As for the ASCII problem, could you add a failing spec for this and I'll work on a fix for it? I reverted my last fix because like I said it broke some specs.

**Comment by luckyraul** [comment]
There was no problem for url with our fix.
